api/patch_user_handler.go

```diff
@@ -1,31 +1,60 @@
 package api
 
 import (
+	"fmt"
+	"net/http"
+
+	"github.com/go-playground/validator"
+
 	"github.com/alphagov/paas-accounts/database"
 	"github.com/labstack/echo"
-	"net/http"
 )
 
+type PatchRequest struct {
+	Email    *string `json:"user_email" validate:"omitempty,email"`
+	Username string  `json:"username" validate:"required"`
+}
+
 func PatchUserHandler(db *database.DB) echo.HandlerFunc {
-		return func(c echo.Context) error {
-			var user database.User
-			err := c.Bind(&user)
-			if err != nil {
-				return err
-			}
+	return func(c echo.Context) error {
+		var payload PatchRequest
+		err := c.Bind(&payload)
+		if err != nil {
+			return err
+		}
 
-			user.UUID = c.Param("uuid")
+		err = c.Validate(payload)
+		if err != nil {
+			valerr := err.(validator.ValidationErrors)
+			s := fmt.Sprint(valerr)
+			return c.JSON(http.StatusBadRequest, s)
+		}
 
-			_, err = db.GetUser(c.Param("uuid"))
-			if err != nil {
+		user, err := db.GetUser(c.Param("uuid"))
+		if err != nil {
+			if err == database.ErrUserNotFound {
 				return c.JSON(http.StatusNotFound, err)
 			}
 
-			err = db.PatchUser(user)
-			if err != nil {
-				return err
-			}
+			return err
+		}
+
+		if user.Username == nil {
+			user.Username = &payload.Username
```

Comment:
This silently fails when `PATCH`ing the username of a user with a username already set. I'd prefer to return an error instead, but open to discussion :)

Reply:
This was deliberate. We figure we don't really ever want usernames to be updated (we think), but we also have a scenario where we have existing users in the database that don't have a username set. Setting the username, but only if it hasn't been set yet, allows us to accommodate that. It's useful when migrating users, and when rolling out the change to new environments.

We'll be introducing a change to paas-accounts deployment for the latter.

Reply:
I decided this was very much personal preference and not a big deal. :)